### PR TITLE
Upgrade to Python 3.12

### PR DESCRIPTION
**Pull request by gnufede**
## Motivation

Move to a Python version with active support

APPSEC-53884

## Changes

TODO list of migration : 

First question : is it still compatible with 3.9? is yes, it could ease the migration.

Then : 

* [ ] DataDog/dd-trace-cpp : https://github.com/DataDog/dd-trace-cpp/blob/main/.circleci/config.yml#L158
* [ ] DataDog/dd-trace-dotnet : https://github.com/DataDog/dd-trace-dotnet/blob/3d66935dcb9ebe7b6421e0e3e502122ebbac98a9/.azure-pipelines/ultimate-pipeline.yml#L4574
* [ ] DataDog/dd-trace-go : https://github.com/DataDog/dd-trace-go/blob/bdc072c35e963da2a370c1afc74cb93c346e7fbd/.github/workflows/system-tests.yml#L111 and https://github.com/DataDog/dd-trace-go/blob/bdc072c35e963da2a370c1afc74cb93c346e7fbd/.github/workflows/parametric-tests.yml#L35
* [x] DataDog/dd-trace-java: https://github.com/DataDog/dd-trace-java/blob/97e2337ea116e9f61c151dc2b54bb4ba09b3b779/.circleci/config.continue.yml.j2#L263. but version pinned, so nothing to do, just call them to update the version when they'll update the commit id.
* [ ] DataDog/dd-trace-js https://github.com/DataDog/dd-trace-js/blob/5daf54de526f2c0477e1eaf27146fb620b88dd03/.github/workflows/system-tests.yml#L35
* [ ] DataDog/dd-trace-php https://github.com/DataDog/dd-trace-php/blob/44a07ba98de2062986a33cc1592a2b87533441a0/.circleci/continue_config.yml#L3495
* [ ] DataDog/dd-trace-py https://github.com/DataDog/dd-trace-py/blob/9c9b5a7e9d5977438cb98a6d86cf00b23f7fa3ff/.github/workflows/system-tests.yml#L27
* [ ] DataDog/dd-trace-rb https://github.com/DataDog/dd-trace-rb/blob/07e3a5a74ad8634a2986cba2c14575afd1e2f8db/.github/workflows/system-tests.yml#L35
* [ ] DataDog/appsec-event-rules https://github.com/DataDog/appsec-event-rules/blob/3f886ebad091d66a48badb64e1e06ff8ee079add/.github/workflows/system-tests.yaml#L62

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

File: docs/execute/requirements.md [hunk 1]
```diff
@@ -22,8 +22,8 @@ TODO
 For Homebrew users :
 
 ```
-brew install python@3.9
-pip3.9 install virtualenv
+brew install python@3.11
+pip3.12 install virtualenv
```
Comment (gnufede) [inline docs/execute/requirements.md:26]:
```suggestion
pip3.11 install virtualenv
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -10,14 +10,17 @@ jobs:
       - checkout
 
       - run:
-          name: Install python 3.9
-          command: sudo apt-get install python3.9-venv
+          name: Install python 3.11
+          command: |
+            sudo add-apt-repository ppa:deadsnakes/ppa
```
Comment (lucaspimentel) [inline .circleci/config.yml:15]: Is adding this repository required?

File: docs/scenarios/parametric.md [hunk 3]
```diff
@@ -39,7 +39,7 @@ Make sure you're in the root of the repository before running these commands.
 The following dependencies are required to run the tests locally:
 
 - Docker
-- Python >= 3.7 (for Windows users, Python 3.9 seems to run best without issues)
+- Python >= 3.7 (for Windows users, Python 3.11 seems to run best without issues)
```
Comment (lucaspimentel) [inline docs/scenarios/parametric.md:42]: Should this be updated?
```suggestion
- Python 3.11
```

File: utils/build/docker/python/parametric/Dockerfile [hunk 4]
```diff
@@ -1,8 +1,8 @@
 
 FROM ghcr.io/datadog/dd-trace-py/testrunner:7ce49bd78b0d510766fc5db12756a8840724febc
 WORKDIR /app
-RUN pyenv global 3.9.11
-RUN python3.9 -m pip install fastapi==0.89.1 uvicorn==0.20.0
+RUN pyenv global 3.12
+RUN python3.12 -m pip install fastapi==0.89.1 uvicorn==0.20.0
```
Comment (lucaspimentel) [inline utils/build/docker/python/parametric/Dockerfile:5]: Why `3.12` here instead of `3.11`?

File: .circleci/config.yml (hunk 2 above)
Comment (gnufede) [reply .circleci/config.yml:15]: Yes, as 3.11 is not available at least in this CircleCI runner (not sure where we can pick a more updated runner)

File: utils/build/docker/python/parametric/Dockerfile (hunk 4 above)
Comment (gnufede) [inline utils/build/docker/python/parametric/Dockerfile:5]:
```suggestion
RUN pyenv global 3.11
RUN python3.11 -m pip install fastapi==0.89.1 uvicorn==0.20.0
```

File: lib-injection/build/docker/python/dd-lib-python-init-test-django-gunicorn/Dockerfile [hunk 5]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9
+FROM python:3.11
```
Comment (cbeauchesne) [inline lib-injection/build/docker/python/dd-lib-python-init-test-django-gunicorn/Dockerfile:1]: No need for this one, it's a tested app. I mean, maybe python team will want to test python3.11, but it's not related with this current change.

Comment (cbeauchesne) [reply lib-injection/build/docker/python/dd-lib-python-init-test-django-gunicorn/Dockerfile:1]: and same for any other dockerfile

Comment (gnufede) [inline lib-injection/build/docker/python/dd-lib-python-init-test-django-gunicorn/Dockerfile:1]:
```suggestion
FROM python:3.9
```

File: lib-injection/build/docker/python/dd-lib-python-init-test-django-uvicorn/Dockerfile [hunk 6]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9
+FROM python:3.11
```
Comment (gnufede) [inline lib-injection/build/docker/python/dd-lib-python-init-test-django-uvicorn/Dockerfile:1]:
```suggestion
FROM python:3.9
```

File: lib-injection/build/docker/python/dd-lib-python-init-test-django/Dockerfile [hunk 7]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9
+FROM python:3.11
```
Comment (gnufede) [inline lib-injection/build/docker/python/dd-lib-python-init-test-django/Dockerfile:1]:
```suggestion
FROM python:3.9
```

File: utils/build/docker/python/django-poc.base.Dockerfile [hunk 8]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9-slim
+FROM python:3.11-slim
```
Comment (gnufede) [inline utils/build/docker/python/django-poc.base.Dockerfile:1]:
```suggestion
FROM python:3.9-slim
```

File: utils/build/docker/python/flask-poc.base.Dockerfile [hunk 9]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9-slim
+FROM python:3.11-slim
```
Comment (gnufede) [inline utils/build/docker/python/flask-poc.base.Dockerfile:1]:
```suggestion
FROM python:3.9-slim
```

File: utils/build/docker/python/parametric/ddtracer_version.Dockerfile [hunk 10]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9-slim
+FROM python:3.11-slim
```
Comment (gnufede) [inline utils/build/docker/python/parametric/ddtracer_version.Dockerfile:1]:
```suggestion
FROM python:3.9-slim
```

File: utils/build/docker/python/uwsgi-poc.base.Dockerfile [hunk 11]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.9-slim
+FROM python:3.11-slim
```
Comment (gnufede) [inline utils/build/docker/python/uwsgi-poc.base.Dockerfile:1]:
```suggestion
FROM python:3.9-slim
```

File: lib-injection/build/docker/python/dd-lib-python-init-test-django-gunicorn/Dockerfile (hunk 5 above)
Comment (gnufede) [reply lib-injection/build/docker/python/dd-lib-python-init-test-django-gunicorn/Dockerfile]: Maybe except `runner`? And `interfaces/schemas`?

File: docs/execute/requirements.md [hunk 12]
```diff
@@ -22,8 +22,8 @@ TODO
 For Homebrew users :
 
 ```
-brew install python@3.9
-pip3.9 install virtualenv
+brew install python@3.12
+pip3.11 install virtualenv
```
Comment (gnufede) [inline docs/execute/requirements.md:26]:
```suggestion
pip3.12 install virtualenv
```

File: tests/parametric/conftest.py [hunk 13]
```diff
@@ -108,13 +108,13 @@ def python_library_factory() -> APMLibraryTestServer:
         container_img="""
 FROM ghcr.io/datadog/dd-trace-py/testrunner:7ce49bd78b0d510766fc5db12756a8840724febc
```
Comment (gnufede) [inline tests/parametric/conftest.py:109]: We will need to change this once after the new `testrunner` image is built and uploaded, so code below would work

File: .circleci/config.yml [hunk 14]
```diff
@@ -4,20 +4,21 @@ jobs:
...
     resource_class: large
     steps:
       - checkout
 
       - run:
-          name: Install python 3.9
-          command: sudo apt-get install python3.9-venv
+          name: Install python 3.12
+          command: |
+            sudo apt-get install python3.12-venv python3.12-dev
```
Comment (gnufede) [inline .circleci/config.yml:15]:
```suggestion
            sudo apt-get install python3.12-venv
```

Comment (gnufede) [inline .circleci/config.yml:15]:
```suggestion
          command: sudo apt-get install python3.12-venv
```

File: tests/parametric/conftest.py (hunk 13 above)
Comment (cbeauchesne) [reply tests/parametric/conftest.py:109]: which testrunner image ?

**Review by cbeauchesne — COMMENTED** [review]
Seems good to me. Asked a review from @robertomonteromiguel  also.

Once it's ready, we need to create Pr on other repo to update their CI too.

Let's plan this next week

**Comment by cbeauchesne** [comment]
It'll probably break local install, to reduce complaints, could you add  on run.sh a detection if it's python 3.9, and exit with an error message : 

> Please run `rm -rf venv && ./build.sh -i runner` 🙏

**Comment by cbeauchesne** [comment]
ok, LGTM. Let me list repos where we need to do a PR. I'll put the list in the first message

**Review by cbeauchesne — APPROVED** [review]

**Pull request merged**